src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,1305 @@
+        if( part.removed ) {
+            continue;
+        }
+        tripoint pos;
+        driven_veh.coord_translate( tdir, pivot, part.mount, pos );
+        if( extent_map.find( pos.y ) == extent_map.end() ) {
+            extent_map[pos.y] = { pos.x, pos.x };
+        } else {
+            auto &extent = extent_map[pos.y];
+            extent.first = std::min( extent.first, pos.x );
+            extent.second = std::max( extent.second, pos.x );
+        }
+    }
+    for( const auto &extent : extent_map ) {
+        const int y = extent.first;
+        const int x_min = extent.second.first;
+        const int x_max = extent.second.second;
+        for( int x = x_min; x <= x_max; x++ ) {
+            ret.occupied_zone.emplace_back( x, y );
+        }
+    }
+    for( int part_num : driven_veh.rotors ) {
+        const vehicle_part &part = driven_veh.part( part_num );
+        const int diameter = part.info().rotor_diameter();
+        const int radius = ( diameter + 1 ) / 2;
+        if( radius > 0 ) {
+            tripoint pos;
+            driven_veh.coord_translate( tdir, pivot, part.mount, pos );
+            for( tripoint pt : points_in_radius( pos, radius ) ) {
+                ret.occupied_zone.emplace_back( pt.xy() );
+            }
+        }
+    }
+    // figure out the maximum amount of displacement that can happen from advancing the
+    // tileray a single step by advancing it a bunch of times and reducing the x and y
+    // components of the displacement to at most 1
+    constexpr int far_advance = 100;
+    tdir.advance( far_advance );
+    const point increment( signum( tdir.dx() ), signum( tdir.dy() ) );
+    // compute the subset of vehicle points that would move to previously unoccupied spots
+    // (a.k.a. "moving first") when the vehicle moves one step; these are the only points
+    // we need to check for collision when the vehicle is moving in this direction
+    const std::unordered_set<point> occupied_set( ret.occupied_zone.begin(), ret.occupied_zone.end() );
+    for( const point &pt : ret.occupied_zone ) {
+        if( occupied_set.find( pt + increment ) == occupied_set.end() ) {
+            ret.collision_points.emplace_back( pt );
+        }
+    }
+    return ret;
+}
+
+
+// Return true if the map tile at the given position (in map coordinates)
+// can be driven on (not an obstacle).
+bool vehicle::autodrive_controller::check_drivable( tripoint pt ) const
+{
+    const map &here = get_map();
+
+    // check if another vehicle is there; tiles occupied by the current
+    // vehicle are evidently drivable
+    const optional_vpart_position ovp = here.veh_at( pt );
+    if( ovp ) {
+        // Known corner case: some furniture can be driven over, but will collide with
+        // wheel parts; if the vehicle starts over such a furniture we'll mark that tile
+        // as safe and may collide with it by turning; however if we mark it unsafe
+        // we'll have no viable paths away from the starting point.
```

Comment:
Future improvement: higher cost instead of just a yes/no return from this function. This could also be used for "padding" of various types - around monsters, etc.

Reply:
I see what you're saying, but the boolean is still needed. No matter how high a cost you assign to obstacles, A* will still choose to go through them if there are no other paths.

So maybe return a tuple of bool and integer (is it drivable and what's the cost), to be stored in separate arrays.